[LLDB] Add framework for Data Inspection Language (DIL) work.

cmtice opened the pull request:
Add the framework code for hooking up and calling the Data Inspection Language (DIL) implementation, as an alternate implementation for the 'frame variable' command. For now, this is an opt-in option, via a target setting 'target.experimental.use-DIL'.  See
https://discourse.llvm.org/t/rfc-data-inspection-language/69893 for more information about this project.

This PR does not actually call any of the DIL code; instead the piece that will eventually call the DIL code (StackFrame::DILEvaluateVariableExpression) calls back into the original 'frame variable' implementation.

cmtice committed:
[LLDB] Add framework for Data Inspection Language (DIL) work.

Add the framework code for hooking up and calling the Data Inspection Language
(DIL) implementation, as an alternate implementation for the 'frame
variable' command. For now, this is an opt-in option, via a target setting
'target.experimental.use-DIL'.  See
https://discourse.llvm.org/t/rfc-data-inspection-language/69893 for more
information about this project.

This PR does not actually call any of the DIL code; instead the piece that will
eventually call the DIL code (StackFrame::DILEvaluateVariableExpression) calls
back into the original 'frame variable' implementation.

jimingham commented:
IIUC, the goal is that In the end, DIL will win and there will only be one VariableExpressionPath interface lldb.  So it seems inefficient for the interim stage to introduce another function and switch between them with UseDIL all over the code.  Why not leave GetValueForVariableExpressionPath for now, and move the `if (UseDIL())` into that function.  They have the same signature, and you'll always be able to find the target from the ValueObject when you are in GetValueForVariableExpressionPath, so that should be possible.

That will make the diff much smaller, and in the end mean that pulling the switch to use only the DIL will be less disruptive.

labath commented on lldb/source/Plugins/SymbolFile/DWARF/DWARFASTParser.cpp:
```diff
@@ -56,8 +56,19 @@ DWARFASTParser::ParseChildArrayInfo(const DWARFDIE &parent_die,
-                      var_die.GetName(), eNoDynamicValues, 0, var_sp, error);
+                  lldb::TargetSP target_sp = frame->CalculateTarget();
+                  bool use_DIL = target_sp->GetUseDIL(
+                      (lldb_private::ExecutionContext *)exe_ctx);
```
It looks like this cast wouldn't be required if you changed `GetExperimentalPropertyValue` to take a const execution context (all of the functions below it already take a const argument)

cmtice replied:
This code & cast are gone now.

labath commented:
+1 to what Jim said. You could keep `GetValueForVariableExpressionPath` as a single entry point, but then implement it as something like `if(use_dil) LegacyGetValueForVariableExpressionPath() else DILGetValueForVariableExpressionPath()`

cmtice committed:
Clean up DIL framework.

    Remove redundant calculations of 'use_DIL'. Make
    StackFrame::GetValueForVariableExpressionPath the gating function
    that dispatches to either DILGetValueForVariableExpressionPath or
    LegacyGetValueForVariableExpressionPath.

cmtice committed:
Restore SBFrame.cpp to original code.

cmtice commented:
I've updated the code as Jim and Pavel suggested. :-)

labath commented on lldb/source/Target/StackFrame.cpp:
```diff
@@ -508,6 +508,32 @@ StackFrame::GetInScopeVariableList(bool get_file_globals,
+  if (use_DIL)
+    return DILGetValueForVariableExpressionPath(var_expr, use_dynamic, options,
+                                                var_sp, error);
+  else
```
https://llvm.org/docs/CodingStandards.html#don-t-use-else-after-a-return

(I also think the comments in this method don't tell us anything that's not immediately obvious from the code. I'd just remove them.)

cmtice replied:
Done.

labath commented on lldb/source/Target/StackFrame.cpp:
```diff
@@ -508,6 +508,32 @@ StackFrame::GetInScopeVariableList(bool get_file_globals,
     llvm::StringRef var_expr, DynamicValueType use_dynamic, uint32_t options,
     VariableSP &var_sp, Status &error) {
+  // Check to see if we should use the DIL implementation.
+  lldb::TargetSP target_sp = CalculateTarget();
+  ExecutionContext exe_ctx;
+  CalculateExecutionContext(exe_ctx);
+  bool use_DIL = target_sp->GetUseDIL(&exe_ctx);
```
```suggestion
  ExecutionContext exe_ctx;
  CalculateExecutionContext(exe_ctx);
  bool use_DIL = exe_ctx.GetTargetRef().GetUseDIL(&exe_ctx);
```

cmtice replied:
Done.

werat commented on lldb/source/Target/Target.cpp:
```diff
@@ -4385,6 +4385,27 @@ bool TargetProperties::GetInjectLocalVariables(
+    return exp_values->GetPropertyAtIndexAs<bool>(ePropertyUseDIL, exe_ctx)
+        .value_or(false);
+  else
+    return true;
```
Should this be `false` as well? Otherwise DIL will be enabled if `exp_values` is not available (not sure when this happens though).

cmtice committed:
Minor cleanups.

- Make LegacyGetValueForVariableExpressionPath and
  DILGetValueForVariableExpressionPath private.
- Remove redundant or useless comments.
- Clean up definition of use_DIL slightly.

cmtice commented on lldb/include/lldb/Target/StackFrame.h:
```diff
@@ -304,6 +304,42 @@ class StackFrame : public ExecutionContextScope,
+  ///
+  /// \return
+  ///     A shared pointer to the ValueObject described by var_expr.
+  lldb::ValueObjectSP DILGetValueForVariableExpressionPath(
```
Done.
